File: apps/api/src/lib/category-budget-status.ts
/**
 * List-with-status computation for category budgets, extracted from the GET /
 * handler in routes/category-budgets.ts. Resolves each budget to a specific
 * month, computes actual spending (net of refunds, split-aware) and the
 * pay-period effective-expected figure. Read-only — no ledger mutations.
 */
import { prisma } from '@budget-tracker/db';
import type { z } from 'zod';
import type { BudgetStatusResponseSchema } from '@budget-tracker/core';
import { resolveEffectiveVersion, computeBudgetStatus, isSeasonalActiveInMonth } from './budget.js';
import { localDate, today } from './dates.js';
import { generateSchedule } from './schedule-generator.js';
import { prorateBudget } from './spend-prediction.js';
import { serializeCategoryBudget, budgetInclude } from './category-budget-serialization.js';

export interface ListBudgetStatusesQuery {
  yearPlanId: string;
  month?: number;
  year?: number;
  includeSeasonal: boolean;
  periodStart?: Date;
  periodEnd?: Date;
  viewMode?: 'PAY_PERIOD' | 'MONTHLY' | 'ANNUAL';
}

type BudgetStatusResponse = z.infer<typeof BudgetStatusResponseSchema>;

/**
 * Compute the status list for a year plan's budgets. Returns `null` when the
 * year plan does not exist.
 */
export async function listBudgetStatuses(
  query: ListBudgetStatusesQuery,
): Promise<BudgetStatusResponse[] | null> {
  const now = today();
  const nd = localDate(now);
  const month = query.month ?? nd.month + 1;
  const year = query.year ?? nd.year;

  const yearPlan = await prisma.yearPlan.findUnique({ where: { id: query.yearPlanId } });
  if (!yearPlan) return null;

  const budgets = await prisma.categoryBudget.findMany({
    where: { yearPlanId: query.yearPlanId, removedAt: null },
    include: budgetInclude,
  });

  const monthStart = new Date(Date.UTC(year, month - 1, 1));
  const monthEnd = new Date(Date.UTC(year, month, 1));

  // Use pay period dates if provided, otherwise calendar month
  const spendStart = query.periodStart ?? monthStart;
  const spendEnd = query.periodEnd ?? monthEnd;
  const isPeriodMode = !!(query.periodStart && query.periodEnd) && query.viewMode !== 'ANNUAL';

  // In pay period mode, compute recurring expense totals per category
  // using the same logic as the spend prediction
  const recurringByCategory = new Map<string, number>();
  let scheduleType = 'BIWEEKLY';
  if (isPeriodMode) {
    await generateSchedule({ periodStart: spendStart, periodEnd: spendEnd });

    // Get the pay schedule type for proration
    const schedule =
      (await prisma.paySchedule.findFirst({ where: { isDefault: true } })) ??
      (await prisma.paySchedule.findFirst({ orderBy: { createdAt: 'asc' } }));
    if (schedule) scheduleType = schedule.type;

    // Sum recurring expenses per category from ScheduledTransaction rows
    const scheduleRows = await prisma.scheduledTransaction.findMany({
      where: {
        dueDate: { gte: spendStart, lte: spendEnd },
        sourceType: 'EXPENSE',
      },
      include: { expense: { select: { id: true, budgetId: true } } },
    });
    for (const row of scheduleRows) {
      if (!row.expense?.budgetId) continue;
      const catId = row.expense.budgetId;
      recurringByCategory.set(
        catId,
        (recurringByCategory.get(catId) ?? 0) + Number(row.expectedAmount),
      );
    }
  }

  // Batch query: actual spending per category for the period
  // Net spending = expenses minus refunds tagged to the same budget
  // Split transaction handling:
  //   - Non-split transactions (no children): count full amount toward their budgetId
  //   - Child transactions: count their amount toward their budgetId
  //   - Split parents (have children): count (parent.amount - sum(children.amount)) toward parent's budgetId
  //     This "remainder" is the portion the user kept on the original category (row 1 in the split modal)
  const categoryIds = budgets.map((b) => b.budgetId);
  const spendingByCategory = new Map<string, number>();
  if (categoryIds.length > 0) {
    // 1. Sum all non-parent transactions (unsplit + children) per budgetId
    const [spendingRows, refundRows] = await Promise.all([
      prisma.transaction.groupBy({
        by: ['budgetId'],
        _sum: { amount: true },
        where: {
          type: 'EXPENSE',
          budgetId: { in: categoryIds },
          date: { gte: spendStart, lt: spendEnd },
          NOT: { children: { some: {} } },
        },
      }),
      prisma.transaction.groupBy({
        by: ['budgetId'],
        _sum: { amount: true },
        where: {
          type: 'REFUND',
          budgetId: { in: categoryIds },
          date: { gte: spendStart, lt: spendEnd },
          NOT: { children: { some: {} } },
        },
      }),
    ]);
    for (const row of spendingRows) {
      spendingByCategory.set(row.budgetId!, row._sum.amount?.toNumber() ?? 0);
    }
    for (const row of refundRows) {
      const current = spendingByCategory.get(row.budgetId!) ?? 0;
      spendingByCategory.set(row.budgetId!, current - (row._sum.amount?.toNumber() ?? 0));
    }

    // 2. For split parents: add the remainder (parent.amount - sum(children.amount)) to parent's budgetId
    const splitParents = await prisma.transaction.findMany({
      where: {
        type: 'EXPENSE',
        budgetId: { in: categoryIds },
        date: { gte: spendStart, lt: spendEnd },
        children: { some: {} },
      },
      select: { id: true, amount: true, budgetId: true, children: { select: { amount: true } } },
    });
    for (const parent of splitParents) {
      const childSum = parent.children.reduce((sum, ch) => sum + Number(ch.amount), 0);
      const remainder = Number(parent.amount) - childSum;
      if (remainder > 0 && parent.budgetId) {
        const current = spendingByCategory.get(parent.budgetId) ?? 0;
        spendingByCategory.set(parent.budgetId, current + remainder);
      }
    }
  }

  const results: BudgetStatusResponse[] = [];

  for (const budget of budgets) {
    const effectiveVersion = resolveEffectiveVersion(budget.versions, month, year);
    if (
      !query.includeSeasonal &&
      effectiveVersion &&
      effectiveVersion.activeMonths.length > 0 &&
      !isSeasonalActiveInMonth(effectiveVersion.activeMonths, month)
    )
      continue;

    const actualSpending = spendingByCategory.get(budget.budgetId) ?? 0;

    // In pay period mode, determine effectiveExpected based on whether the budget
    // has linked recurring expenses:
    // - Linked expenses exist AND some are due: max(recurring, prorated budget)
    // - Linked expenses exist but none due this period: $0
    // - No linked expenses (discretionary): prorated budget
    let effectiveExpected: number | undefined;
    if (isPeriodMode && effectiveVersion) {
      const recurringTotal = recurringByCategory.get(budget.budgetId) ?? 0;
      const hasLinkedExpenses = budget._count.budgetExpenseLinks > 0;
      const monthlyBudget = effectiveVersion.monthlyEquivalent.toNumber();
      const activeMonths =
        effectiveVersion.activeMonths.length > 0 ? effectiveVersion.activeMonths : null;
      const proratedBudget = prorateBudget(
        monthlyBudget,
        'MONTHLY',
        activeMonths,
        scheduleType,
        spendStart,
        spendEnd,
      );

      if (hasLinkedExpenses) {
        // Mandatory: driven by recurring schedule
        if (recurringTotal > 0) {
          effectiveExpected = Math.max(recurringTotal, proratedBudget);
        } else {
          // No recurring expenses due this period — still show prorated budget
          effectiveExpected = proratedBudget;
        }
      } else {
        // Discretionary: prorated budget applies every period
        effectiveExpected = proratedBudget;
      }
    }

    // Determine if this seasonal budget is inactive for the current month
    const isInactiveSeasonal =
      effectiveVersion &&
      effectiveVersion.activeMonths.length > 0 &&
      !isSeasonalActiveInMonth(effectiveVersion.activeMonths, month);

    /*
     * Seasonal zeroing is a statement about THIS MONTH, so it has no business in
     * the annual view. A summer-only budget is genuinely zero in December —
     * looking at December. Looking at the YEAR it had a real allocation and real
     * spending, and zeroing drops it out of the annual totals entirely. Same for
     * `doneForYear`: finished is not the same as never budgeted.
     */
    const isAnnual = query.viewMode === 'ANNUAL';
    const isZeroed = !isAnnual && (isInactiveSeasonal || budget.doneForYear);

    /*
     * The ANNUAL view compares a YEAR of spending, so it compares against a YEAR
     * of budget — x12, matching `convertToFrequency(monthly, 'ANNUAL')` in the
     * web app. Without it the status badge measured months of spending against
     * one month's allocation and read "over" on nearly everything, while the
     * card beside it drew its progress bar against x12: two numbers on one card
     * disagreeing about what period they described.
     */
    const monthly = effectiveVersion ? effectiveVersion.monthlyEquivalent.toNumber() : 0;
    // A seasonal budget has (active months) of annual allowance, not twelve —
    // judging three months of budget against a year calls it wildly under-spent.
    const annualMonths =
      effectiveVersion && effectiveVersion.activeMonths.length > 0
        ? effectiveVersion.activeMonths.length
        : 12;
    const compareAmount = isZeroed
      ? 0
      : (effectiveExpected ?? (isAnnual ? monthly * annualMonths : monthly));
    const status = effectiveVersion ? computeBudgetStatus(actualSpending, compareAmount) : null;

    const serialized = serializeCategoryBudget(budget, effectiveVersion);
    // Zero out budget amounts for inactive seasonal or done-for-year budgets so totals are correct
    if (isZeroed && serialized.version) {
      serialized.version.monthlyEquivalent = 0;
    }
    const finalEffectiveExpected = isZeroed ? 0 : effectiveExpected;

    results.push({
      ...serialized,
      actualSpending,
      effectiveExpected: finalEffectiveExpected,
      status,
    });
  }

  return results;
}

File: rust/api/src/category_budget_status.rs
//! The three fields `GET /category-budgets` adds to each allocation.
//!
//! Ported from `apps/api/src/lib/category-budget-status.ts`, which step 4 missed
//! entirely. The route returned every budget in a well-formed array without
//! `actualSpending` or `status`, both required by `BudgetStatusResponseSchema` —
//! so Zod threw in the browser and the Budgets page rendered empty while the
//! endpoint answered 200 and every Rust test passed.
//!
//! # Spending is split-aware, and that is the whole subtlety
//!
//! A split purchase is one parent with N children. Counting both would double
//! it. The rule:
//!
//! - **Unsplit rows and children** count in full, against their own `budgetId`.
//! - **A split parent** counts only its *remainder* — `parent.amount −
//!   SUM(children.amount)` — against the parent's budget. That remainder is the
//!   portion the user kept on the original category rather than reassigning.
//!
//! Refunds subtract, so a returned purchase stops counting as spending.

use avoir_core::budget::{compute_budget_status, is_seasonal_active_in_month};
use avoir_core::money::Cents;
use avoir_core::prediction::{prorate_budget, BudgetPeriod};
use chrono::NaiveDate;
use serde_json::Value;
use sqlx::SqliteConnection;
use std::collections::HashMap;

use crate::ApiError;

/// The window and mode a status list is computed for.
pub(crate) struct Window {
    pub start: String,
    /// Exclusive upper bound, matching the TypeScript's `lt`.
    pub end_exclusive: String,
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
}

/// Net spending per budget for the window.
pub(crate) async fn spending_by_budget(
    conn: &mut SqliteConnection,
    w: &Window,
) -> Result<HashMap<String, Cents>, ApiError> {
    let mut out: HashMap<String, Cents> = HashMap::new();

    // Rows with no children of their own: unsplit purchases and the children
    // themselves. Both count in full.
    let rows = sqlx::query!(
        r#"SELECT t."budgetId" AS "budget_id!", t."type" AS "ty!",
                  SUM(t."amount") AS "total!: i64"
             FROM "Transaction" t
            WHERE t."budgetId" IS NOT NULL
              AND t."type" IN ('EXPENSE', 'REFUND')
              AND t."date" >= ?1 AND t."date" < ?2
              AND NOT EXISTS (SELECT 1 FROM "Transaction" c WHERE c."parentId" = t."id")
            GROUP BY t."budgetId", t."type""#,
        w.start,
        w.end_exclusive
    )
    .fetch_all(&mut *conn)
    .await?;

    for r in rows {
        let slot = out.entry(r.budget_id).or_insert(Cents(0));
        if r.ty == "REFUND" {
            *slot -= Cents(r.total);
        } else {
            *slot += Cents(r.total);
        }
    }

    // Split parents contribute only what was not reassigned to a child.
    let parents = sqlx::query!(
        r#"SELECT t."budgetId" AS "budget_id!", t."amount" AS "amount!: i64",
                  COALESCE((SELECT SUM(c."amount") FROM "Transaction" c
                             WHERE c."parentId" = t."id"), 0) AS "children!: i64"
             FROM "Transaction" t
            WHERE t."budgetId" IS NOT NULL AND t."type" = 'EXPENSE'
              AND t."date" >= ?1 AND t."date" < ?2
              AND EXISTS (SELECT 1 FROM "Transaction" c WHERE c."parentId" = t."id")"#,
        w.start,
        w.end_exclusive
    )
    .fetch_all(&mut *conn)
    .await?;

    for p in parents {
        let remainder = p.amount - p.children;
        // Only a positive remainder counts. A parent whose children exceed it is
        // over-allocated — adding a negative here would credit the category with
        // spending that never happened.
        if remainder > 0 {
            *out.entry(p.budget_id).or_insert(Cents(0)) += Cents(remainder);
        }
    }

    Ok(out)
}

/// Recurring expense totals per budget, from the period's occurrences.
pub(crate) async fn recurring_by_budget(
    conn: &mut SqliteConnection,
    w: &Window,
) -> Result<HashMap<String, Cents>, ApiError> {
    let rows = sqlx::query!(
        r#"SELECT e."budgetId" AS "budget_id!",
                  SUM(s."expectedAmount") AS "total!: i64"
             FROM "ScheduledTransaction" s
             JOIN "Expense" e ON e."id" = s."expenseId"
            WHERE s."sourceType" = 'EXPENSE'
              AND s."dueDate" >= ?1 AND s."dueDate" <= ?2
            GROUP BY e."budgetId""#,
        w.start,
        w.end_exclusive
    )
    .fetch_all(&mut *conn)
    .await?;

    Ok(rows
        .into_iter()
        .map(|r| (r.budget_id, Cents(r.total)))
        .collect())
}

/// What one budget is expected to absorb this pay period.
///
/// Only meaningful in period mode; a monthly view compares against the monthly
/// equivalent directly.
///
/// - **With linked recurring expenses**, the floor is the prorated allocation
///   but the bills win when they exceed it: a period holding a £900 rent payment
///   against a £400 prorated share is expected to spend £900, not £400.
/// - **Without them** the budget is discretionary, and the prorated share is the
///   whole story.
pub(crate) fn effective_expected(
    monthly_equivalent: Cents,
    active_months: &[u32],
    has_linked_expenses: bool,
    recurring_total: Cents,
    schedule_type: &str,
    w: &Window,
) -> Cents {
    let prorated = prorate_budget(
        monthly_equivalent,
        BudgetPeriod::Monthly,
        if active_months.is_empty() {
            None
        } else {
            Some(active_months)
        },
        schedule_type,
        w.start_date,
        w.end_date,
    );
    if has_linked_expenses && recurring_total.0 > prorated.0 {
        recurring_total
    } else {
        prorated
    }
}

/// Attach `actualSpending`, `effectiveExpected` and `status` to one serialized
/// allocation, and zero it out when it should not count this month.
///
/// A budget that is done for the year, or seasonal and out of season, compares
/// against **zero** — and its `monthlyEquivalent` is zeroed in the response too,
/// so the page's totals do not include an allocation nobody can spend.
#[allow(clippy::too_many_arguments)]
pub(crate) fn decorate(
    mut serialized: Value,
    actual: Cents,
    monthly_equivalent: Cents,
    active_months: &[u32],
    month: u32,
    done_for_year: bool,
    period_mode: bool,
    expected: Option<Cents>,
    has_version: bool,
    // The ANNUAL view compares a YEAR of spending, so it must compare against a
    // YEAR of budget. Without this the status badge measured months of spending
    // against one month's allocation and read "over" on nearly everything, while
    // the card beside it drew its progress bar against x12 — two numbers on one
    // card disagreeing about what period they described.
    annual: bool,
) -> Value {
    /*
     * Seasonal zeroing is a statement about THIS MONTH, so it has no business in
     * the annual view.
     *
     * A summer-only budget is genuinely zero in December — looking at December,
     * there is nothing to spend. Looking at the YEAR, it had a real allocation
     * and real spending, and reporting it as zero drops it out of the annual
     * totals entirely. The same applies to `doneForYear`: finished is not the
     * same as never budgeted, and the year is exactly the period over which it
     * WAS budgeted.
     */
    let inactive_seasonal =
        !active_months.is_empty() && !is_seasonal_active_in_month(active_months, month);
    let zeroed = !annual && (inactive_seasonal || done_for_year);

    let compare = if zeroed {
        Cents(0)
    } else if period_mode {
        expected.unwrap_or(monthly_equivalent)
    } else if annual {
        /*
         * A YEAR of budget. For an ordinary budget that is x12, matching
         * `convertToFrequency(monthly, 'ANNUAL')` in the web app — the frontend
         * draws the bar and the backend judges the status, and a card showing
         * both is one card.
         *
         * For a SEASONAL budget it is x(active months): a budget that only runs
         * three months of the year has three months of annual allowance, and
         * judging it against twelve would call every seasonal budget wildly
         * under-spent.
         */
        let months = if active_months.is_empty() {
            12
        } else {
            active_months.len() as i64
        };
        Cents(monthly_equivalent.0 * months)
    } else {
        monthly_equivalent
    };

    // No version means no allocation was ever set, which is not the same as an
    // allocation of zero — the first has no status, the second is `None` from
    // `compute_budget_status` anyway.
    let status = if has_version {
        compute_budget_status(actual, compare)
    } else {
        None
    };

    if zeroed {
        if let Some(v) = serialized.get_mut("version") {
            if let Some(obj) = v.as_object_mut() {
                obj.insert("monthlyEquivalent".into(), serde_json::json!(0.0));
            }
        }
    }

    let obj = serialized
        .as_object_mut()
        .expect("a serialized budget is an object");
    obj.insert(
        "actualSpending".into(),
        serde_json::json!(actual.as_dollars_f64()),
    );
    obj.insert(
        "status".into(),
        match status {
            Some(s) => serde_json::json!(s),
            None => Value::Null,
        },
    );
    if period_mode {
        let e = if zeroed {
            Cents(0)
        } else {
            expected.unwrap_or(Cents(0))
        };
        obj.insert(
            "effectiveExpected".into(),
            serde_json::json!(e.as_dollars_f64()),
        );
    }
    serialized
}
